### PR TITLE
Parse function and object types in Flow

### LINKED ISSUE #738: Parse function and object types in Flow
Contributes to #584 
- [x]  function type
- [x]  object type, optional properties in object type, exact object type, indexer property for object type (aka map object)

### PR DESCRIPTION
**Pull request by inverno**
Fixes #738

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 1]
```diff
@@ -1750,6 +1754,66 @@ public FlowLiteralTypeTree flowLiteralType(InternalSyntaxToken token) {
...
+    }
+
+    return new SeparatedListImpl<>(parameters, commas);
+  }
+
+  public FlowFunctionTypeParameterTree flowFunctionTypeParameter(IdentifierTree identifierTree, FlowTypeAnnotationTree typeAnnotationTree) {
+    return new FlowFunctionTypeParameterTreeImpl();
+  }
+
+  public FlowFunctionTypeParameterTree flowFunctionTypeParameter(FlowTypeTree type) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: MAJOR Remove this unused method parameter "type". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 2]
```diff
@@ -1750,6 +1754,66 @@ public FlowLiteralTypeTree flowLiteralType(InternalSyntaxToken token) {
     return new FlowLiteralTypeTreeImpl(null, token);
   }
 
+  public FlowFunctionTypeTree flowFunctionType(FlowFunctionTypeParameterClauseTree flowFunctionTypeParameterClauseTree, InternalSyntaxToken doubleArrow,
+    FlowTypeTree flowTypeTree) {
+    return new FlowFunctionTypeTreeImpl();
+  }
+
+  public FlowFunctionTypeParameterClauseTree flowFunctionTypeParameterClause(
+    InternalSyntaxToken lParenthesis,
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: MAJOR Remove this unused method parameter "lParenthesis". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 3]
```diff
@@ -1750,6 +1754,66 @@ public FlowLiteralTypeTree flowLiteralType(InternalSyntaxToken token) {
...
+    SeparatedList<FlowFunctionTypeParameterTree> parameters,
+    InternalSyntaxToken comma,
+    FlowFunctionTypeRestParameterTree restParameter,
+    InternalSyntaxToken rParenthesis) {
+    return new FlowFunctionTypeParameterClauseTreeImpl();
+  }
+
+  public FlowFunctionTypeParameterClauseTree flowFunctionTypeParameterClause(
+    InternalSyntaxToken lParenthesis,
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: MAJOR Remove this unused method parameter "lParenthesis". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 4]
```diff
@@ -1750,6 +1754,66 @@ public FlowLiteralTypeTree flowLiteralType(InternalSyntaxToken token) {
...
+
+  public FlowFunctionTypeParameterTree flowFunctionTypeParameter(IdentifierTree identifierTree, FlowTypeAnnotationTree typeAnnotationTree) {
+    return new FlowFunctionTypeParameterTreeImpl();
+  }
+
+  public FlowFunctionTypeParameterTree flowFunctionTypeParameter(FlowTypeTree type) {
+    return new FlowFunctionTypeParameterTreeImpl();
+  }
+
+  public FlowFunctionTypeRestParameterTree flowFunctionTypeRestParameter(InternalSyntaxToken token, FlowFunctionTypeParameterTree flowFunctionTypeParameterTree) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: MAJOR Remove this unused method parameter "token". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 5]
```diff
@@ -1750,6 +1754,66 @@ public FlowLiteralTypeTree flowLiteralType(InternalSyntaxToken token) {
...
+  public FlowFunctionTypeParameterClauseTree flowFunctionTypeParameterClause(
+    InternalSyntaxToken lParenthesis,
+    SeparatedList<FlowFunctionTypeParameterTree> parameters,
+    Optional<InternalSyntaxToken> comma,
+    InternalSyntaxToken rParenthesis) {
+    return new FlowFunctionTypeParameterClauseTreeImpl();
+  }
+
+  public FlowFunctionTypeParameterClauseTree flowFunctionTypeParameterClause(
+    InternalSyntaxToken lParenthesis,
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: MAJOR Remove this unused method parameter "lParenthesis". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 6]
```diff
@@ -1750,6 +1754,66 @@ public FlowLiteralTypeTree flowLiteralType(InternalSyntaxToken token) {
     return new FlowLiteralTypeTreeImpl(null, token);
   }
 
+  public FlowFunctionTypeTree flowFunctionType(FlowFunctionTypeParameterClauseTree flowFunctionTypeParameterClauseTree, InternalSyntaxToken doubleArrow,
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: MAJOR Remove this unused method parameter "flowFunctionTypeParameterClauseTree". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java [hunk 7]
```diff
@@ -1750,6 +1754,66 @@ public FlowLiteralTypeTree flowLiteralType(InternalSyntaxToken token) {
...
+      for (Tuple<InternalSyntaxToken, FlowFunctionTypeParameterTree> t : trailingParameters.get()) {
+        commas.add(t.first());
+        parameters.add(t.second());
+      }
+    }
+
+    return new SeparatedListImpl<>(parameters, commas);
+  }
+
+  public FlowFunctionTypeParameterTree flowFunctionTypeParameter(IdentifierTree identifierTree, FlowTypeAnnotationTree typeAnnotationTree) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/TreeFactory.java]: MAJOR Remove this unused method parameter "identifierTree". [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1172)

**Comment by SonarTech** [comment]
SonarQube analysis reported 7 issues
* ![MAJOR][MAJOR] 7 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowFunctionTypeParameterClauseTree.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+import org.sonar.plugins.javascript.api.tree.Tree;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+@Beta
+public interface FlowFunctionTypeParameterClauseTree extends Tree {
+
+  SyntaxToken leftParenthesis();
+
+  @Nullable
+  SeparatedList<FlowFunctionTypeParameterTree> parameters();
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowFunctionTypeParameterClauseTree.java]: @inverno it should not be null, it should be empty

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowFunctionTypeParameterClauseTreeImpl.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  }
+
+  @Override
+  public Kind getKind() {
+    return Kind.FLOW_FUNCTION_TYPE_PARAMETER_CLAUSE;
+  }
+
+  @Override
+  public Iterator<Tree> childrenIterator() {
+    if (parameters != null) {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowFunctionTypeParameterClauseTreeImpl.java]: @inverno when you will make `parameters` not nullable, you can simplify this method

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowFunctionTypeParameterClauseTree.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+  @Nullable
+  SyntaxToken trailingComma();
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowFunctionTypeParameterClauseTree.java]: @inverno if rest parameter is there, this comma is not trailing

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowFunctionTypeRestParameterTreeImpl.java [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+import org.sonar.plugins.javascript.api.tree.Tree;
+import org.sonar.plugins.javascript.api.tree.expression.IdentifierTree;
+import org.sonar.plugins.javascript.api.tree.flow.FlowFunctionTypeParameterTree;
+import org.sonar.plugins.javascript.api.tree.flow.FlowFunctionTypeRestParameterTree;
+import org.sonar.plugins.javascript.api.tree.flow.FlowTypeAnnotationTree;
+import org.sonar.plugins.javascript.api.tree.flow.FlowTypeTree;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitor;
+
+public class FlowFunctionTypeRestParameterTreeImpl extends JavaScriptTree implements FlowFunctionTypeRestParameterTree {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowFunctionTypeRestParameterTreeImpl.java]: @inverno let's drop this interface and class

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowFunctionTypeTree.java [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.javascript.api.tree.flow;
+
+import com.google.common.annotations.Beta;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+@Beta
+public interface FlowFunctionTypeTree extends FlowTypeTree {
+  SyntaxToken doubleArrowToken();
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowFunctionTypeTree.java]: @inverno reorder

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java [hunk 13]
```diff
@@ -622,4 +566,16 @@ public void visitFlowTypedBindingElement(FlowTypedBindingElementTree tree) {
   public void visitFlowOptionalBindingElement(FlowOptionalBindingElementTree tree) {
     scanChildren(tree);
   }
+
+  public void visitFlowFunctionType(FlowFunctionTypeTree tree) {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java]: @inverno move up

File: javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/flow/FlowFunctionTypeParameterClauseTreeModelTest.java [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+
+public class FlowFunctionTypeParameterClauseTreeModelTest extends JavaScriptTreeModelTest {
+
+  @Test
+  public void one_parameter() throws Exception {
+    FlowFunctionTypeParameterClauseTree tree = parse("(par1: string)", Tree.Kind.FLOW_FUNCTION_TYPE_PARAMETER_CLAUSE, Tree.Kind.FLOW_FUNCTION_TYPE_PARAMETER_CLAUSE);
+
+    assertThat(tree.is(Tree.Kind.FLOW_FUNCTION_TYPE_PARAMETER_CLAUSE)).isTrue();
+    assertThat(tree.parameters().size()).isEqualTo(1);
+    assertThat(tree.childrenStream().filter(Objects::nonNull).count()).isEqualTo(3);
```
Comment (vilchik-elena) [inline javascript-frontend/src/test/java/org/sonar/javascript/tree/impl/flow/FlowFunctionTypeParameterClauseTreeModelTest.java]: @inverno replace with DoubleDispatchVisitorTest

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowSimplePropertyTypeKeyTree.java [hunk 15]
```diff
@@ -0,0 +1,33 @@
+package org.sonar.plugins.javascript.api.tree.flow;/*
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowSimplePropertyTypeKeyTree.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowIndexerPropertyTypeKeyTree.java [hunk 16]
```diff
@@ -0,0 +1,33 @@
+package org.sonar.plugins.javascript.api.tree.flow;/*
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowIndexerPropertyTypeKeyTree.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowPropertyTypeTree.java [hunk 17]
```diff
@@ -0,0 +1,27 @@
+package org.sonar.plugins.javascript.api.tree.flow;/*
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowPropertyTypeTree.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowPropertyTypeKeyTree.java [hunk 18]
```diff
@@ -0,0 +1,26 @@
+package org.sonar.plugins.javascript.api.tree.flow;/*
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowPropertyTypeKeyTree.java]: MINOR Move this trailing comment on the previous empty line. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3ATrailingCommentCheck)

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowObjectTypeTreeImpl.java [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitor;
+
+public class FlowObjectTypeTreeImpl extends JavaScriptTree implements FlowObjectTypeTree {
+  private final SyntaxToken lcurlyToken;
+  private final SyntaxToken lpipeToken;
+  private final SeparatedList<Tree> properties;
+  private final SyntaxToken rpipeToken;
+  private final SyntaxToken rcurlyToken;
+
+  public FlowObjectTypeTreeImpl(SyntaxToken lcurlyToken, @Nullable SyntaxToken lpipeToken, SeparatedList<Tree> properties, @Nullable SyntaxToken rpipeToken, SyntaxToken rcurlyToken) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowObjectTypeTreeImpl.java]: MINOR Split this 183 characters long line (which is greater than 180 authorized). [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS00103)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* ![MINOR][MINOR] 5 minor

Watch the comments in this conversation to review them.

[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java [hunk 20]
```diff
@@ -1800,6 +1804,44 @@ public FlowFunctionTypeParameterTree FLOW_FUNCTION_TYPE_REST_PARAMETER() {
...
+    return b.<FlowPropertyTypeTree>nonterminal(Kind.FLOW_PROPERTY_TYPE)
+      .is(b.firstOf(
+        f.flowPropertyType(IDENTIFIER_NAME(), b.optional(b.token(JavaScriptPunctuator.QUERY)), FLOW_TYPE_ANNOTATION()),
+        f.flowPropertyType(FLOW_INDEXER_PROPERTY_TYPE_KEY(), FLOW_TYPE_ANNOTATION()))
+      );
+  }
+
+  public FlowIndexerPropertyTypeKeyTree FLOW_INDEXER_PROPERTY_TYPE_KEY() {
+    return b.<FlowIndexerPropertyTypeKeyTree>nonterminal()
+      .is(f.flowIndexerPropertyTypeKey(b.token(JavaScriptPunctuator.LBRACKET), IDENTIFIER_NAME(), b.token(JavaScriptPunctuator.RBRACKET)));
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @inverno you lost this syntax:
```
var obj: { [user_id: number]: string } = {};
```

Comment (vilchik-elena) [reply javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @inverno and it should be not identifier but type

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowIndexerPropertyTypeKeyTree.java (hunk 16 above)
Comment (vilchik-elena) [reply javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowIndexerPropertyTypeKeyTree.java]: @inverno -_- something is wrong here

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowPropertyTypeTree.java (hunk 17 above)
Comment (vilchik-elena) [reply javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/flow/FlowPropertyTypeTree.java]: @inverno same thing

File: javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java [hunk 21]
```diff
@@ -1800,6 +1804,44 @@ public FlowFunctionTypeParameterTree FLOW_FUNCTION_TYPE_REST_PARAMETER() {
...
+          b.token(JavaScriptPunctuator.RCURLYBRACE)
+        )
+      ));
+  }
+
+  public SeparatedList<Tree> FLOW_OBJECT_TYPE_PROPERTIES() {
+    return b.<SeparatedList<Tree>>nonterminal()
+      .is(f.properties(
+        FLOW_PROPERTY_TYPE(),
+        b.zeroOrMore(f.newTuple(b.token(JavaScriptPunctuator.COMMA), FLOW_PROPERTY_TYPE())),
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/parser/JavaScriptGrammar.java]: @inverno support `;`

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowPropertyTypeTreeImpl.java [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+  private final FlowTypeAnnotationTree typeAnnotation;
+
+  public FlowPropertyTypeTreeImpl(FlowPropertyTypeKeyTree key, FlowTypeAnnotationTree typeAnnotation) {
+    this.key = key;
+    this.typeAnnotation = typeAnnotation;
+  }
+
+  @Override
+  public Kind getKind() {
+    return Kind.FLOW_PROPERTY_TYPE;
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowPropertyTypeTreeImpl.java]: rename to FLOW_PROPERTY_DEFINITION

File: javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowPropertyTypeTreeImpl.java [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+import com.google.common.collect.Iterators;
+import java.util.Iterator;
+import org.sonar.javascript.tree.impl.JavaScriptTree;
+import org.sonar.plugins.javascript.api.tree.Tree;
+import org.sonar.plugins.javascript.api.tree.flow.FlowPropertyTypeKeyTree;
+import org.sonar.plugins.javascript.api.tree.flow.FlowPropertyTypeTree;
+import org.sonar.plugins.javascript.api.tree.flow.FlowTypeAnnotationTree;
+import org.sonar.plugins.javascript.api.visitors.DoubleDispatchVisitor;
+
+public class FlowPropertyTypeTreeImpl extends JavaScriptTree implements FlowPropertyTypeTree {
```
Comment (vilchik-elena) [inline javascript-frontend/src/main/java/org/sonar/javascript/tree/impl/flow/FlowPropertyTypeTreeImpl.java]: FlowPropertyDefinition

**Pull request merged**